packages/react/src/hooks/useGetUrl.ts

```diff
@@ -0,0 +1,68 @@
+import * as React from 'react';
+
+import { getUrl, GetUrlInput } from 'aws-amplify/storage';
+
+import { isFunction } from '@aws-amplify/ui';
+import { useHasValueUpdated } from '@aws-amplify/ui-react-core';
+
+export type UseGetUrlInput = GetUrlInput & {
+  onError?: (error: Error) => void;
+};
+
+interface UseGetUrlOutput {
+  isLoading: boolean;
+  url: URL | undefined;
+  expiresAt: Date | undefined;
+}
+
+function getKeyOrPath(
+  input: UseGetUrlInput
+): string | (({ identityId }: { identityId?: string }) => string) {
+  if (input.key !== undefined) {
+    return input.key;
+  }
+  return input.path;
+}
+
+export const useGetUrl = (input: UseGetUrlInput): UseGetUrlOutput => {
+  const [result, setResult] = React.useState<UseGetUrlOutput>({
+    url: undefined,
+    expiresAt: undefined,
+    isLoading: false,
+  });
+  const keyOrPath = getKeyOrPath(input);
+  const hasImgUpdated = useHasValueUpdated(keyOrPath);
+
+  React.useEffect(() => {
+    if (!hasImgUpdated) {
+      return;
+    }
+    const { onError, ...getUrlInput } = input;
+    setResult((prevResult) => ({ ...prevResult, isLoading: true }));
+    let ignore = false;
+
+    getUrl(getUrlInput)
+      .then((response) => {
+        if (ignore) {
+          return;
+        }
+        setResult((prevResult) => ({ ...prevResult, ...response }));
+      })
+      .catch((error: Error) => {
+        if (ignore) {
+          return;
+        }
+        if (isFunction(onError)) {
+          onError(error);
+        }
+
+        return () => {
```

Comment:
A lot of this logic is reused from useStorageURL. Believe this line is meant to prevent unnecessary re-renders so if the component unmounts before resolving the url the state doesn't get set. https://github.com/aws-amplify/amplify-ui/blob/22f8c98607a9e77e767ffbe22fca174f54bd3381/packages/react/src/hooks/useStorageURL.ts#L53

Reply:
In this case we do want it to be returned conditionally though right? Because it should only run if the promise fails to resolve?

Reply:
Looked at the original hook and saw that cleanup effect is returned from the `catch`, which is incorrect as well. The `ignore` should ensure that only the results of the most recent call are returned, and should always be returned unconditionally